nexus/src/db/collection_detach.rs

```diff
@@ -0,0 +1,1117 @@
+        pool: &db::Pool,
+    ) -> Collection {
+        let create_params = IdentityMetadataCreateParams {
+            name: Name::try_from(name.to_string()).unwrap(),
+            description: "description".to_string(),
+        };
+        let c =
+            Collection { identity: CollectionIdentity::new(id, create_params) };
+
+        diesel::insert_into(collection::table)
+            .values(c)
+            .execute_async(pool.pool())
+            .await
+            .unwrap();
+
+        get_collection(id, &pool).await
+    }
+
+    async fn get_collection(id: Uuid, pool: &db::Pool) -> Collection {
+        collection::table
+            .find(id)
+            .select(Collection::as_select())
+            .first_async(pool.pool())
+            .await
+            .unwrap()
+    }
+
+    async fn insert_resource(
+        id: Uuid,
+        name: &str,
+        pool: &db::Pool,
+    ) -> Resource {
+        let create_params = IdentityMetadataCreateParams {
+            name: Name::try_from(name.to_string()).unwrap(),
+            description: "description".to_string(),
+        };
+        let r = Resource {
+            identity: ResourceIdentity::new(id, create_params),
+            collection_id: None,
+        };
+
+        diesel::insert_into(resource::table)
+            .values(r)
+            .execute_async(pool.pool())
+            .await
+            .unwrap();
+
+        get_resource(id, &pool).await
+    }
+
+    async fn attach_resource(
+        collection_id: Uuid,
+        resource_id: Uuid,
+        pool: &db::Pool,
+    ) {
+        Collection::attach_resource(
+            collection_id,
+            resource_id,
+            collection::table.into_boxed(),
+            resource::table.into_boxed(),
+            100,
+            diesel::update(resource::table)
+                .set(resource::dsl::collection_id.eq(collection_id)),
+        )
+        .attach_and_get_result_async(pool.pool())
+        .await
+        .unwrap();
+    }
+
+    async fn get_resource(id: Uuid, pool: &db::Pool) -> Resource {
+        resource::table
+            .find(id)
+            .select(Resource::as_select())
+            .first_async(pool.pool())
+            .await
+            .unwrap()
+    }
+
+    #[test]
+    fn test_verify_query() {
+        let collection_id =
+            uuid::Uuid::parse_str("cccccccc-cccc-cccc-cccc-cccccccccccc")
+                .unwrap();
+        let resource_id =
+            uuid::Uuid::parse_str("aaaaaaaa-aaaa-aaaa-aaaa-aaaaaaaaaaaa")
+                .unwrap();
+        let detach = Collection::detach_resource(
+            collection_id,
+            resource_id,
+            collection::table.into_boxed(),
+            resource::table.into_boxed(),
+            diesel::update(resource::table)
+                .set(resource::dsl::collection_id.eq(Option::<Uuid>::None)),
+        );
+        let query = diesel::debug_query::<Pg, _>(&detach).to_string();
+
+        let expected_query = "WITH \
```

Comment:
the attach and detach CTEs are very similar, it's a shame one can't be implemented by calling the other. for example, detach could call attach's CTE and supply `resource_count = SELECT 1` and `max_items = 2` to short-circuit attach's max items check.

Reply:
Forget this comment :) I don't think it's worth it, because attach and detach won't always remain similar mirror images of each other.